test/test_xla_backend.py

```diff
@@ -0,0 +1,362 @@
+def new_group_barrier_disabled():
+  orig_barrier_fn = torch.distributed.distributed_c10d._store_based_barrier
+  torch.distributed.distributed_c10d._store_based_barrier = lambda x, y, z: None
+  try:
+    yield
+  finally:
+    torch.distributed.distributed_c10d._store_based_barrier = orig_barrier_fn
+
+
+@contextmanager
+def xm_cc_op_intercepted(cc_op):
+  orig_xm_cc_op = getattr(xm, cc_op)
+
+  def build_collective_context(groups):
+    cctx = xm.CollectiveContext()
+    # Here we need to manipulate cctx to force xm.all_reduce to generate
+    # the collective communication ops in HLO.
+    # We may need to .requires_interhost_reduce in future.
+    cctx.requires_intercore_reduce = True
+    cctx.intercore_group = groups
+    return cctx
+
+  def patched_cc_op(*args, **kwargs):
+    kwargs['cctx'] = build_collective_context(kwargs['groups'])
+    return orig_xm_cc_op(*args, **kwargs)
+
+  setattr(xm, cc_op, patched_cc_op)
+  try:
+    yield
+  finally:
+    setattr(xm, cc_op, orig_xm_cc_op)
+
+
+class XlaBackendTest(unittest.TestCase):
+
+  @classmethod
+  def setUpClass(cls):
+    dist.init_process_group(
+        'xla', rank=0, world_size=1, init_method='tcp://localhost:6789')
+
+  def test_xla_backend_exists(self):
+    # torch_xla.distributed._register_xla_backend() should have been
+    # automatically called.
+    pg_xla_creator = dist.Backend.XLA
+    assert pg_xla_creator is not None
+
+  def test_process_group_creation(self):
+    pg_xla = get_process_group_xla(rank=1, size=2)
+    assert pg_xla is not None
+    assert pg_xla.rank() == 1
+    assert pg_xla.size() == 2
+
+  def test_allreduce(self):
+    device = xm.xla_device()
+    tensor = torch.arange(2, device=device) + 1 + 2 * dist.get_rank()
+    pg_xla = get_process_group_xla(rank=511, size=1024)
+    opts = dist.AllreduceOptions()
+    opts.reduceOp = dist.ReduceOp.SUM
+    all_reduce_pattern = r'%all\-reduce\.\d+ = .+ all\-reduce\('
+    with xm_cc_op_intercepted('all_reduce'):
+      pg_xla.allreduce([tensor], opts)
+    hlo = torch_xla._XLAC._get_xla_tensors_hlo([tensor])
+    hlo_matches(hlo, all_reduce_pattern)
+    xm.mark_step()  # purge all computations attached the device.
+
+  def test_allreduce_with_mesh(self):
+    device = xm.xla_device()
+    tensor = torch.arange(2, device=device) + 1 + 2 * dist.get_rank()
+
+    set_world_size(6)
+    ranks = [2, 3]
+    world_rank = 3
+    set_world_rank(world_rank)
+    with new_group_barrier_disabled():
+      new_pg = dist.new_group(ranks=ranks)
+    opts = dist.AllreduceOptions()
+    opts.reduceOp = dist.ReduceOp.SUM
+    all_reduce_pattern = (r'%all\-reduce\.\d+ = .+ all\-reduce\(.+\), .*'
+                          r'replica_groups=\{\{0,1\},\{2,3\},\{4,5\}\}')
+    with xm_cc_op_intercepted('all_reduce'):
+      new_pg.allreduce([tensor], opts)
+    hlo = torch_xla._XLAC._get_xla_tensors_hlo([tensor])
+    hlo_matches(hlo, all_reduce_pattern)
+    xm.mark_step()  # purge all computations attached the device.
```

Comment:
really nit thing, can we move the comment above the `mark_step` instead of after the `mark_step`? This matches our python styles on other tests.

Reply:
Done